src/box/replication.cc

```diff
@@ -1055,8 +1055,11 @@ replicaset_connect(const struct uri_set *uris,
 	}
 	int count = 0;
 	struct applier *appliers[VCLOCK_MAX] = {};
+	/* Memory for on_state triggers registered in appliers */
+	struct applier_on_connect triggers[VCLOCK_MAX];
 	auto appliers_guard = make_scoped_guard([&]{
 		for (int i = 0; i < count; i++) {
+			trigger_clear(&triggers[i].base);
```

Comment:
If `applier_new()` a few lines below outside of the guard will raise an exception, then the guard will try to clear triggers which were not created yet, i.e. they contain garbage pointers right now.

Reply:
Yes, thanks for noticing!

I tried rearranging the code so that `trigger_clear` works correctly in this guard, but it turned out too bulky:
<details><summary>diff attempt</summary>

```diff
diff --git a/src/box/replication.cc b/src/box/replication.cc
index 76d4eee78..7db8b1f24 100644
--- a/src/box/replication.cc
+++ b/src/box/replication.cc
@@ -1065,23 +1065,6 @@ replicaset_connect(const struct uri_set *uris,
 			applier_delete(applier);
 		}
 	});
-	for (; count < uris->uri_count; count++)
-		appliers[count] = applier_new(&uris->uris[count]);
-
-	say_info("connecting to %d replicas", count);
-
-	if (!connect_quorum) {
-		/*
-		 * Enter orphan mode on configuration change and
-		 *
-		 * only leave it when we manage to sync with
-		 * replicaset_quorum instances. Don't change
-		 * title though, it should be 'loading' during
-		 * local recovery.
-		 */
-		box_do_set_orphan(true);
-	}
-
 	/*
 	 * Simultaneously connect to remote peers to receive their UUIDs
 	 * and fill the resulting set:
@@ -1101,21 +1084,36 @@ replicaset_connect(const struct uri_set *uris,
 	state.connected = state.failed = 0;
 	fiber_cond_create(&state.wakeup);
 
-	double timeout = replication_connect_timeout;
-
-	/* Add triggers and start simulations connection to remote peers */
-	for (int i = 0; i < count; i++) {
-		struct applier *applier = appliers[i];
-		struct applier_on_connect *trigger = &triggers[i];
+	for (; count < uris->uri_count; count++) {
+		appliers[count] = applier_new(&uris->uris[count]);
+		struct applier_on_connect *trigger = &triggers[count];
 		/* Register a trigger to wake us up when peer is connected */
 		trigger_create(&trigger->base, applier_on_connect_f, NULL, NULL);
 		trigger->state = &state;
 		trigger->seen_state = APPLIER_CONNECT;
-		trigger_add(&applier->on_state, &trigger->base);
-		/* Start background connection */
-		applier_start(applier);
+		trigger_add(&appliers[count]->on_state, &trigger->base);
 	}
 
+	say_info("connecting to %d replicas", count);
+
+	if (!connect_quorum) {
+		/*
+		 * Enter orphan mode on configuration change and
+		 *
+		 * only leave it when we manage to sync with
+		 * replicaset_quorum instances. Don't change
+		 * title though, it should be 'loading' during
+		 * local recovery.
+		 */
+		box_do_set_orphan(true);
+	}
+
+	double timeout = replication_connect_timeout;
+
+	/* Start simulations connection to remote peers */
+	for (int i = 0; i < count; i++)
+		applier_start(appliers[i]);
+
 	while (!replicaset_is_connected(&state, appliers, count,
 					connect_quorum)) {
 		double wait_start = ev_monotonic_now(loop());
```
</details>

So instead I use 2 guards now. I tried to avoid that in initial patch, and reuse the applier guard, but it's not ok to clear an uninitialized trigger, as you pointed out.
<details><summary>incremental diff</summary>

```diff
diff --git a/src/box/replication.cc b/src/box/replication.cc
index 76d4eee78..feb36c957 100644
--- a/src/box/replication.cc
+++ b/src/box/replication.cc
@@ -1055,11 +1055,8 @@ replicaset_connect(const struct uri_set *uris,
        }
        int count = 0;
        struct applier *appliers[VCLOCK_MAX] = {};
-       /* Memory for on_state triggers registered in appliers */
-       struct applier_on_connect triggers[VCLOCK_MAX];
        auto appliers_guard = make_scoped_guard([&]{
                for (int i = 0; i < count; i++) {
-                       trigger_clear(&triggers[i].base);
                        struct applier *applier = appliers[i];
                        applier_stop(applier);
                        applier_delete(applier);
@@ -1097,6 +1094,9 @@ replicaset_connect(const struct uri_set *uris,
         *   for duplicates, fill the result set, return.
         */
 
+       /* Memory for on_state triggers registered in appliers */
+       struct applier_on_connect triggers[VCLOCK_MAX];
+
        struct replicaset_connect_state state;
        state.connected = state.failed = 0;
        fiber_cond_create(&state.wakeup);
@@ -1116,6 +1116,16 @@ replicaset_connect(const struct uri_set *uris,
                applier_start(applier);
        }
 
+       /*
+        * Guards are run in the order reverse to the order of their
+        * initialization. So triggers_guard will run before appliers_guard.
+        */
+       auto triggers_guard = make_scoped_guard([&]{
+               for (int i = 0; i < count; i++) {
+                       struct applier_on_connect *trigger = &triggers[i];
+                       trigger_clear(&trigger->base);
+               }
+       });
        while (!replicaset_is_connected(&state, appliers, count,
                                        connect_quorum)) {
                double wait_start = ev_monotonic_now(loop());
@@ -1148,6 +1158,7 @@ replicaset_connect(const struct uri_set *uris,
                if (applier->state != APPLIER_CONNECTED)
                        applier_stop(applier);
        }
+       triggers_guard.is_active = false;
 
        /* Now all the appliers are connected, update the replica set. */
        replicaset_update(appliers, count, keep_connect);
```
</details>

Reply:
An alternative was to create them with `NULL` functions right after the declaration. And then destroy + create with the actual callback later. So the triggers would never be in an invalid state. But I am fine with the present solution too.